src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,551 @@
+#define SLOW_AZ_C2		Q_CONVERT_FLOAT(0.98, 15)
+
+/* Threshold for notifying user space, no more often than every 100 ms */
+#define CONTROL_UPDATE_MIN_TIME	Q_CONVERT_FLOAT(100e-3, 16)
+
+/* Emphasis filter for sound direction */
+static const uint32_t iir_emphasis_48k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xd7b49168, 0x62c27f22,
+	0x19712b75, 0xcd1da916, 0x19712b75, 0x00000000,
+	0x00007fb2
+};
+
+static const uint32_t iir_emphasis_16k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xef646e59, 0x2cc88203,
+	0x0fb60984, 0xe093ecf7, 0x0fb60984, 0x00000000,
+	0x00007fb2
+};
+
+/* Function to copy data from component source to cross correlation buffer with emphasis
+ * filtering. Skip if not tracking.
+ */
+void tdfb_direction_copy_emphasis(struct tdfb_comp_data *cd, int ch_count, int *ch, int32_t x)
+{
+	int32_t y;
+
+	if (!cd->direction_updates)
+		return;
+
+	y = iir_df2t(&cd->direction.emphasis[*ch], x);
+	*cd->direction.wp = sat_int16(Q_SHIFT_RND(y, 31, 17)); /* 12 dB boost after high-pass */
+	cd->direction.wp++;
+	tdfb_cinc_s16(&cd->direction.wp, cd->direction.d_end, cd->direction.d_size);
+	(*ch)++;
+	if (*ch == ch_count)
+		*ch = 0;
+}
+
+/* Use function y = sqrt_int16(x), where x is Q4.12 unsigned, y is Q4.12
+ * TODO: Add scaling of input to Q4.12 with a suitable N^2 or (1/N)^2 value
+ * to increase sqrt() computation precision and avoid MIN() function to
+ * saturate for large distances (over 4m). The current virtual source
+ * distance 3m allows max 1m size array for this to work correctly without
+ * hitting the saturation.
+ */
+static inline int16_t tdfb_mic_distance_sqrt(int32_t x)
+{
+	int32_t xs;
+
+	xs = Q_SHIFT_RND(x, 24, 12);
+	xs = MIN(xs, UINT16_MAX);
+	return sqrt_int16((uint16_t)xs);
+}
+
+static inline int16_t max_mic_distance(struct tdfb_comp_data *cd)
+{
+	int16_t d;
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int32_t d2;
+	int32_t d2_max = 0;
+	int i;
+	int j;
+
+	/* Max lag based on largest array dimension. Microphone coordinates are Q4.12 meters */
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		for (j = 0; i < cd->config->num_mic_locations; i++) {
+			if (j == i)
+				continue;
+
+			dx = cd->mic_locations[i].x - cd->mic_locations[j].x;
+			dy = cd->mic_locations[i].y - cd->mic_locations[j].y;
+			dz = cd->mic_locations[i].z - cd->mic_locations[j].z;
+
+			/* d2 is Q8.24 meters */
+			d2 = dx * dx + dy * dy + dz * dz;
+			if (d2 > d2_max)
+				d2_max = d2;
+		}
+	}
+
+	/* Squared distance is Q8.24, return Q4.12 meters */
+	d = tdfb_mic_distance_sqrt(d2_max);
+	return d;
+}
+
+static inline bool line_array_mode_check(struct tdfb_comp_data *cd)
+{
+	int32_t px, py, pz;
+	int16_t a, b, c, d, e, f;
+	int i;
+	int num_mic_locations = cd->config->num_mic_locations;
+
+	if (num_mic_locations == 2)
+		return true;
+
+	/* Cross product of vectors AB and AC is (0, 0, 0) if they are co-linear.
```

Comment:
possible candidate for generic math lib implementation?

Reply:
Yep, I moved the cross product equation to an inline function sof/math/numbers.h. I was hesitating to move more from coordinates related part to lib because it could be rather specific to this application.